### ISSUE TITLE
Failure when deploying second stack with AWS interactive provider

### ISSUE THREAD

**Issue by danielkza** [remind101/stacker#196]
I get a traceback when attempting to deploy two stacks (second depends on the first) with the new interactive option (but *not* without it). The first stack deploys successfully, but the second doesn't even start.

Running from master at 3604de9e4e67ebd572941ce9ccc3b2e25cb28908.

```
[2016-08-21T04:20:32] Using Interactive AWS Provider
[2016-08-21T04:20:32] Plan Status:
[2016-08-21T04:20:32] 	cobli-auth: pending
[2016-08-21T04:20:32] 	cobli-vpn:  pending
[2016-08-21T04:20:32] Tailing stack: cobli-auth
[2016-08-21T04:20:44] [cobli-auth] CREATE_IN_PROGRESS AWS::S3::Bucket
[2016-08-21T04:20:44] [cobli-auth] CREATE_IN_PROGRESS AWS::S3::Bucket Resource creation Initiated
[2016-08-21T04:21:05] [cobli-auth] CREATE_COMPLETE AWS::S3::Bucket
[2016-08-21T04:21:10] [cobli-auth] CREATE_IN_PROGRESS AWS::IAM::ManagedPolicy
[2016-08-21T04:21:10] [cobli-auth] CREATE_IN_PROGRESS AWS::IAM::Policy
[2016-08-21T04:21:10] [cobli-auth] CREATE_IN_PROGRESS AWS::IAM::ManagedPolicy Resource creation Initiated
[2016-08-21T04:21:10] [cobli-auth] CREATE_COMPLETE AWS::IAM::ManagedPolicy
[2016-08-21T04:21:10] [cobli-auth] CREATE_IN_PROGRESS AWS::IAM::ManagedPolicy
[2016-08-21T04:21:10] [cobli-auth] CREATE_IN_PROGRESS AWS::IAM::Policy Resource creation Initiated
[2016-08-21T04:21:10] [cobli-auth] CREATE_COMPLETE AWS::IAM::Policy
[2016-08-21T04:21:10] [cobli-auth] CREATE_IN_PROGRESS AWS::IAM::ManagedPolicy Resource creation Initiated
[2016-08-21T04:21:10] [cobli-auth] CREATE_COMPLETE AWS::IAM::ManagedPolicy
[2016-08-21T04:21:13] Tailing stack: cobli-vpn
Traceback (most recent call last):
  File "/home/danielkza/venv/bin/stacker", line 6, in <module>
    exec(compile(open(__file__).read(), __file__, 'exec'))
  File "/home/danielkza/venv/src/stacker/scripts/stacker", line 9, in <module>
    args.run(args)
  File "/home/danielkza/venv/src/stacker/stacker/commands/stacker/build.py", line 52, in run
    dump=options.dump)
  File "/home/danielkza/venv/src/stacker/stacker/actions/base.py", line 125, in execute
    self.run(*args, **kwargs)
  File "/home/danielkza/venv/src/stacker/stacker/actions/build.py", line 301, in run
    plan.execute()
  File "/home/danielkza/venv/src/stacker/stacker/plan.py", line 251, in execute
    if not self._single_run():
  File "/home/danielkza/venv/src/stacker/stacker/plan.py", line 219, in _single_run
    status = step.run()
  File "/home/danielkza/venv/src/stacker/stacker/plan.py", line 72, in run
    return self._run_func(self.stack, status=self.status)
  File "/home/danielkza/venv/src/stacker/stacker/actions/build.py", line 184, in _launch_stack
    provider_stack = self.provider.get_stack(stack.fqn)
  File "/home/danielkza/venv/src/stacker/stacker/providers/aws/default.py", line 101, in get_stack
    kwargs=dict(StackName=stack_name))['Stacks'][0]
  File "/home/danielkza/venv/src/stacker/stacker/providers/aws/default.py", line 69, in retry_on_throttling
    retry_checker=_throttling_checker)
  File "/home/danielkza/venv/src/stacker/stacker/util.py", line 57, in retry_with_backoff
    return function(*args, **kwargs)
  File "/home/danielkza/venv/lib/python2.7/site-packages/botocore/client.py", line 159, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/danielkza/venv/lib/python2.7/site-packages/botocore/client.py", line 483, in _make_api_call
    operation_model, request_dict)
  File "/home/danielkza/venv/lib/python2.7/site-packages/botocore/endpoint.py", line 117, in make_request
    return self._send_request(request_dict, operation_model)
  File "/home/danielkza/venv/lib/python2.7/site-packages/botocore/endpoint.py", line 146, in _send_request
    success_response, exception):
  File "/home/danielkza/venv/lib/python2.7/site-packages/botocore/endpoint.py", line 219, in _needs_retry
    caught_exception=caught_exception, request_dict=request_dict)
  File "/home/danielkza/venv/lib/python2.7/site-packages/botocore/hooks.py", line 227, in emit
    return self._emit(event_name, kwargs)
  File "/home/danielkza/venv/lib/python2.7/site-packages/botocore/hooks.py", line 210, in _emit
    response = handler(**kwargs)
  File "/home/danielkza/venv/lib/python2.7/site-packages/botocore/retryhandler.py", line 183, in __call__
    if self._checker(attempts, response, caught_exception):
  File "/home/danielkza/venv/lib/python2.7/site-packages/botocore/retryhandler.py", line 251, in __call__
    caught_exception)
  File "/home/danielkza/venv/lib/python2.7/site-packages/botocore/retryhandler.py", line 266, in _should_retry
    return self._checker(attempt_number, response, caught_exception)
  File "/home/danielkza/venv/lib/python2.7/site-packages/botocore/retryhandler.py", line 314, in __call__
    caught_exception)
  File "/home/danielkza/venv/lib/python2.7/site-packages/botocore/retryhandler.py", line 223, in __call__
    attempt_number, caught_exception)
  File "/home/danielkza/venv/lib/python2.7/site-packages/botocore/retryhandler.py", line 356, in _check_caught_exception
    raise caught_exception
ssl.SSLError: [SSL: DECRYPTION_FAILED_OR_BAD_RECORD_MAC] decryption failed or bad record mac (_ssl.c:1754)
```

**Comment by phobologic** [comment]
Can you reproduce this?  It seems like this is an issue w/ SSL & the low level client in botocore, which gets called from the default provider (the interactive provider is a subclass of the default one, but calls many of the same methods, in this case it's calling `providers/aws/default.py", line 69, in retry_on_throttling`).

**Comment by danielkza** [comment]
I can't reproduce it in a different machine (OS X 10.11.6, compared to Fedora 24 from the original report). I'll investigate it further.

**Comment by danielkza** [comment]
This is reproducible in a second Fedora 24 machine. I have a reasonable guess that this is a thread-safety issue with reusing the Boto session in multiple processes, specifically in the background process that waits for the change set to be ready (boto/boto#3170).

**Comment by phobologic** [comment]
Hey @danielkza - can you share the command line options you're using?  Thanks for this!

**Comment by danielkza** [comment]
@phobologic  I believe this doesn't really depend on the options, but on some weird interaction between my Python version, SSL library version and/or Boto. Iḿ just doing `stacker build production config.yml -i --tail`.

**Comment by phobologic** [comment]
Yeah, just want to make sure we have all the info.  I'm not sure if `-i` and `--tail` were programmed to be used together (@mhahn would know, he wrote both) but I don't think that's your issue.

It looks like this specific error is well known with `multiprocessing` when using the `ssl` module:

http://stackoverflow.com/questions/3724900/python-ssl-problem-with-multiprocessing
http://stackoverflow.com/questions/31768631/ssl-error-with-using-multiprocessing-and-sockets

Multiprocessing is only used with `--tail` as far as I know, not the `-i` interactive mode, so at least we can narrow the problem down a bit there.  Can you try it without `--tail` to verify that that is the issue?

**Comment by phobologic** [comment]
Also, I think I have a fix for this - can you give the https://github.com/remind101/stacker/tree/fix_multiprocessing_shared_conn branch a try?  Here's what I changed:

```
diff --git a/stacker/providers/aws/default.py b/stacker/providers/aws/default.py
index 482565b..8dc00d0 100644
--- a/stacker/providers/aws/default.py
+++ b/stacker/providers/aws/default.py
@@ -1,5 +1,6 @@
 import json
 import logging
+import os
 import time

 import boto3
@@ -87,13 +88,15 @@ class Provider(BaseProvider):
     def __init__(self, region, **kwargs):
         self.region = region
         self._outputs = {}
+        self._cloudformation = {}

     @property
     def cloudformation(self):
-        if not hasattr(self, "_cloudformation"):
+        pid = os.getpid()
+        if pid not in self._cloudformation:
             session = boto3.Session(region_name=self.region)
-            self._cloudformation = session.client('cloudformation')
-        return self._cloudformation
+            self._cloudformation[pid] = session.client('cloudformation')
+        return self._cloudformation[pid]

     def get_stack(self, stack_name, **kwargs):
         try:
```

Basically we'll use a different client per process id.

**Comment by danielkza** [comment]
I actually used almost exactly that code as quick fix, but I'd like to make it a little bit prettier for the PR, by adding a `reset_session` method to provider and having it passed as a callback to `Plan`. This way the problem will not surface again if something other than tail is added and it's easier to apply the same fix if clients for other AWS services are used in the future. What do you think?

**Comment by phobologic** [comment]
I'm not sure what exactly reset_session would do (just generate a new session?) and when it would be called by Plan, so hard to say.  I think this keeps things relatively simple.

To be honest, as soon as multiprocessing was suggested, I got nervous because of these sorts of weird interactions.  Now I'm wondering if it's necessary, or if we could poll the tail between updates.

**Comment by danielkza** [comment]
> I'm not sure what exactly reset_session would do (just generate a new session?) 

Set `self._cloudformation` to `None`.

>  and when it would be called by Plan, so hard to say. I think this keeps things relatively simple.

I added a callback parameter to `Plan` to be executed whenever a new process is spawned.

**Comment by phobologic** [comment]
I'd be up for seeing a PR if you have an idea in the works.  Still really wondering if we could simplify multiprocessing out of the picture, but that's probably more work than this needs at the moment.

**Comment by danielkza** [comment]
I think I ended up complicating things a bit too much while trying to properly test with multiprocessing. You can go ahead and merge your fix, since it will be much simpler. It might be wise to leave a comment about it and check if it doesn't behave strangely in testing.

**Comment by danielkza** [comment]
I just tested your branch and it seems to work fine @phobologic. Just a small note looking at the code: it probably isn't necessary to keep a dictionary of sessions, as any single process should never have more than one session. Storing the PID observed when the session was created, and throwing it away if the PID changes should be enough.

**Comment by mhahn** [comment]
I used multiprocessing initially since it was the simplest thing conceptually. We launch multiple stacks at the same time, so when implementing tailing, I thought we could just trigger a `_tail_stack` that was a 1 to 1 mapping with `_launch_stack`. Thus multiprocessing to trigger one process per stack to tail it.

I looked through it again, we could definitely simplify this by just having a single tail function that will loop through all the stacks it should be tailing and output their events. 

I haven't been using tail recently, I'll spend some time thinking about some best practice ways to run stacker and then think about how I could simplify this.
